### PR TITLE
Separate thread for main loop.

### PR DESCRIPTION
**Pull request by denisfa**
Here it is. I will still need to do some cleanup, but at least you can see and check/test how this is going. I will need all commits, despite my description, but I will squash them later.

The good thing here is to no longer pause while browsing menus on Windows, others may include actually being able to maneuver the framerate.

### CODE REVIEW

File: src/wx/panel.cpp [hunk 1]
```diff
@@ -973,6 +961,307 @@ void GameArea::Resume()
...
+                wxGetApp().frame->PollJoysticks();
+                if (!panel) {
+                    wxQueueEvent(this, new wxThreadEvent(WX_THREAD_REQUEST_UPDATEDRAWPANEL));
+                }
+                if (!paused && panel) {
+                    emusys->emuMain(emusys->emuCount);
+                }
+	    }
+        }
+        wxMilliSleep(1);
```
Comment (denisfa) [inline src/wx/panel.cpp:1032]: @rkitover Here we can play a little bit about how to do proper timesteps and other controls. We can try to control when and for how long we should sleep, like an actual main loop for a game.

File: src/wx/panel.cpp [hunk 2]
```diff
@@ -973,6 +961,307 @@ void GameArea::Resume()
...
+    wxQueueEvent(this, event);
+}
+
+struct PrinterDataDialog {
+    uint16_t* to_print;
+    uint16_t** accum_prdata;
+    int lines, feed;
+    int *accum_prdata_len, *accum_prdata_size;
+};
+PrinterDataDialog printerDataDialog;
```
Comment (denisfa) [inline src/wx/panel.cpp:1169]: @rkitover YIKES alert here. If you know something that could help me here with wxWidgets, I would appreciate it very much.

File: src/wx/wxvbam.cpp [hunk 3]
```diff
@@ -772,6 +775,10 @@ void MainFrame::OnDropFile(wxDropFilesEvent& event)
     wxString* f = event.GetFiles();
     // ignore all but last
     wxGetApp().pending_load = f[event.GetNumberOfFiles() - 1];
+
+#ifndef NO_THREAD_MAINLOOP
+    GetPanel()->LoadGame(wxGetApp().pending_load);
+#endif
```
Comment (denisfa) [inline src/wx/wxvbam.cpp:781]: @rkitover If I understood correctly, the IDLE event main loop keeps checking if this is set to load a game and start emulation. I would like to keep this idea away from the thread solution, if possible.

**Comment by rkitover** [comment]
Will start going through this later today, sorry been a rough few days for me.

File: src/wx/wxvbam.h [hunk 4]
```diff
@@ -500,11 +504,42 @@ class DrawingPanelBase;
...
+                 , public wxThreadHelper
+#endif
+{
 public:
     GameArea();
     virtual ~GameArea();
 
+#ifndef NO_THREAD_MAINLOOP
+    virtual wxThread::ExitCode Entry();
+    void StartEmulationThread();
```
Comment (rkitover) [inline src/wx/wxvbam.h:528]: Where are these methods defined? Did you forget to include a file in your commit?

Comment (rkitover) [reply src/wx/wxvbam.h:528]: Sorry I didn't see that some sections of the github diff view need to be expanded.

Comment (rkitover) [reply src/wx/wxvbam.h:528]: No they are actually not there, I misunderstood the point of the expand buttons.

**Comment by denisfa** [comment]
@rkitover Glad to see you are back. This is the skeleton of the changes, I have some more to add here. I will focus on just isolating the separate thread emulation. I should put something this weekend (I have been busy with work).

File: src/wx/panel.cpp [hunk 5]
```diff
@@ -2290,12 +2290,12 @@ void GLDrawingPanel::DrawingPanelInit()
     }
 #elif defined(__WXMSW__)
     typedef char* (*wglext)();
-    wglext wglGetExtensionsString = (wglext)wglGetProcAddress("wglGetExtensionsString");
-    if (wglGetExtensionsString == NULL || strstr(wglGetExtensionsString(), "WGL_EXT_swap_control") == 0) {
-        if (wglGetExtensionsString == NULL)
-            systemScreenMessage(_("No support for wglGetExtensionsString"));
-        else
-            systemScreenMessage(_("No support for WGL_EXT_swap_control"));
+    wglext wglGetExtensionsStringEXT = (wglext)wglGetProcAddress("wglGetExtensionsStringEXT");
```
Comment (rkitover) [inline src/wx/panel.cpp:2584]: I cherry-picked this to master, if you don't mind.

**Pull request closed without merging**